### DISCUSSION TITLE
USB doesn't work after successful boot

### DISCUSSION THREAD

**Discussion by thardie** [zbm-dev/zfsbootmenu#664] [Questions] [answered]
I've got a USB stick connected to the motherboard USB ports. EFI sees it just fine, and I've put zfsbootmenu on there, and it loads fine, and finds my root pool and boot the system fine. Everything works great _except_ the motherboard USB controllers seems to be not seen at all by the newly booted kernel. This means I can't see the USB stick once the system is booted. Unplugging and replugging gets no reaction from the booted system.

I did find another thread that talks about disabling hooks, but this seems like a pretty fundamental issue that should affect more than just me, and I shouldn't have to make weird changes to make this work?

Other than this, the system booted and runs fine. Help?

**Comment by ahesford** [comment]
Some systems need ZBM to tear down USB devices before kexec or those devices will not respond. Many systems don't care. Others seem to require that the tear down not take place. You may be in the last group.

Try booting ZBM with the kernel command-line option

     zbm.skip_hooks=90-xhci-unbind.sh

**Reply by thardie** [reply, marked as answer]
> Some systems need ZBM to tear down USB devices before kexec or those devices will not respond. Many systems don't care. Others seem to require that the tear down not take place. You may be in the last group.
> 
> Try booting ZBM with the kernel command-line option
> 
> ```
>  zbm.skip_hooks=90-xhci-unbind.sh
> ```

Thanks - How would I put that in a config.yaml? I checked the docs for the option, and couldn't find how to translate command line options into config.yaml options.

**Comment by zdykstra** [comment]
Set Kernel.CommandLine - https://docs.zfsbootmenu.org/en/v2.3.x/man/generate-zbm.5.html#kernel , or alternatively, edit the existing EFI bundle with `zbm-kcl`.

**Reply by thardie** [reply]
Is this a kernel command line option or a ZBM option? I thought it was a ZBM option, not a kernel command line option. Sorry if I'm being a noob.

**Reply by classabbyamp** [reply]
it's a kernel commandline option for zbm. zbm reads the kernel commandline it receives to configure certain things, like disabling the usb teardown hook

**Reply by zdykstra** [reply]
ZFSBootMenu is just a linux kernel glued together with a semi-custom initramfs. There are kernel command line options available for it. Everything documented at https://docs.zfsbootmenu.org/en/v2.3.x/man/zfsbootmenu.7.html#command-line-parameters applies to the ZFSBootMenu kernel, and can be set via `Kernel.CommandLine` in `config.yaml` or via `zbm-kcl` for existing EFI files.

**Reply by thardie** [reply]
Oh, that's super useful - Does that mean I could add it to `org.zfsbootmenu:commandline` on the pool and it would work there?

**Reply by ahesford** [reply]
No, that property is read by ZFSBootMenu to set the command-line arguments for the environment it will boot.

**Reply by thardie** [reply]
Thanks again for all your help. I'll try some of the other methods you suggested.

I think my initial confusion was thinking that the `Kernel.CommandLine` was the command line options ZBM would pass to the environment it booted, when it is the kernel command line passed to ZBM itself and nothing to do with the environment it will boot.

I'll try this out during a maintenance tonight and update this once I've got it working

**Reply by thardie** [reply]
Thanks again. This worked great. I used `zbm-kcl` to add the parameters to the EFI boot binary and I'm all good. Thanks again for your help.
